packages/cardano-services/test/TxWorker/utils.test.ts

```diff
@@ -0,0 +1,131 @@
+      if (serviceName === process.env.RABBITMQ_SRV_SERVICE_NAME)
+        return [{ name: 'localhost', port: 5672, priority: 6, weight: 5 }];
+      return [];
+    }
+  }
+}));
+
+describe('TxSubmitWorker abstraction', () => {
+  describe('getRunningTxSubmitWorker', () => {
+    let mockServer: http.Server;
+    let connection: Connection;
+    let txSubmitWorker: RunningTxSubmitWorker;
+    let txSubmitProvider: TxSubmitProvider;
+    const ogmiosPortDefault = 1337;
+    const rabbitmqPortDefault = 5672;
+    const logger = createLogger({ level: 'error', name: 'test' });
+    const dnsResolver = createDnsResolver({ factor: 1.1, maxRetryTime: 1000 }, logger);
+    const srvRecord = { name: 'localhost', port: rabbitmqPortDefault, priority: 1, weight: 1 };
+
+    beforeEach(async () => {
+      connection = Ogmios.createConnectionObject({ port: ogmiosPortDefault });
+      mockServer = createMockOgmiosServer({
+        healthCheck: { response: { networkSynchronization: 0.999, success: true } }
+      });
+      await listenPromise(mockServer, connection);
+      await ogmiosServerReady(connection);
+    });
+
+    afterEach(async () => {
+      if (mockServer !== undefined) {
+        await serverClosePromise(mockServer);
+      }
+
+      await txSubmitWorker.stop();
+    });
+
+    it('should instantiate a running worker with service discovery', async () => {
+      const dnsResolverMock = jest.fn().mockResolvedValueOnce(srvRecord);
+      txSubmitProvider = await getOgmiosTxSubmitProvider(dnsResolver, logger, {
+        cacheTtl: 10_000,
+        ogmiosSrvServiceName: process.env.OGMIOS_SRV_SERVICE_NAME,
+        serviceDiscoveryBackoffFactor: 1.1,
+        serviceDiscoveryTimeout: 1000
+      });
+
+      txSubmitWorker = await getRunningTxSubmitWorker(dnsResolverMock, txSubmitProvider, logger, {
+        cacheTtl: 10_000,
+        rabbitmqSrvServiceName: process.env.RABBITMQ_SRV_SERVICE_NAME,
+        serviceDiscoveryBackoffFactor: 1.1,
+        serviceDiscoveryTimeout: 1000
+      });
+
+      await expect(dnsResolverMock).toBeCalledTimes(1);
+      expect(txSubmitWorker.getStatus()).toEqual('connected');
+    });
+  });
+
+  describe('startTxSubmitWorkerWithDiscovery', () => {
+    it('starts and stops the worker', async () => {
+      const start = jest.fn().mockResolvedValueOnce(void 0);
+      const stop = jest.fn().mockResolvedValueOnce(void 0);
+      const workerFactory = jest.fn().mockImplementation(async () => ({
+        on(_: string, __: any) {},
+        start,
+        stop
+      }));
+
+      const runnableWorker = await startTxSubmitWorkerWithDiscovery(workerFactory);
+      await runnableWorker.stop();
+
+      expect(start).toBeCalledTimes(1);
+      expect(stop).toBeCalledTimes(1);
+      expect(workerFactory).toBeCalledTimes(1);
+    });
+
+    it('start the worker and then fail the listener for connection error events, should create and start a new instance of a running worker', async () => {
```

Comment:
```suggestion
    it('should create and start a new instance of TxWorker when a broker connection error event is received', async () => {
```
I'm a little unsure about how the test works though given there's no use of `CONNECTION_ERROR_EVENT` within the assertion :thinking:

Reply:
We assign `jest.fn()` func to the event listener [here](https://github.com/input-output-hk/cardano-js-sdk/blob/848dc62bb5432762b5939454f75fe62db33aa7e5/packages/cardano-services/test/TxWorker/utils.test.ts#L127) and after that manually invoke it [here](https://github.com/input-output-hk/cardano-js-sdk/blob/848dc62bb5432762b5939454f75fe62db33aa7e5/packages/cardano-services/test/TxWorker/utils.test.ts#L141) which simulates that the [event is emitted](https://github.com/input-output-hk/cardano-js-sdk/blob/39d22556809a465c07e6615c628707ff7ee19c29/packages/cardano-services/src/TxWorker/utils.ts#L38)

Reply:
How about renaming `failListener` as `failWithEvent(CONNECTION_ERROR_EVENT)`, even if it's just to better communicate this simulation?

Reply:
Let's go with this instead, since the argument isn't used, so just creates misdirection

``` ts
let simulateConnectionError: any = jest.fn();

on(_: string, listener: any) {
          simulateConnectionError = listener;
 },

simulateConnectionError();
```

Reply:
https://github.com/input-output-hk/cardano-js-sdk/pull/344/commits/89a1a0d3cf617df8746f6493e70edfa219226835